src/libstd/sys/unix/stack_overflow.rs

```diff
@@ -169,13 +169,27 @@ mod imp {
     }
 
     pub unsafe fn make_handler() -> Handler {
-        let stack = get_stack();
-        sigaltstack(&stack, ptr::null_mut());
-        Handler { _data: stack.ss_sp as *mut libc::c_void }
+        let mut stack = mem::zeroed();
+        sigaltstack(ptr::null(), &mut stack);
+        if stack.ss_flags & SS_DISABLE != 0 {
+            stack = get_stack();
+            sigaltstack(&stack, ptr::null_mut());
+            Handler { _data: stack.ss_sp as *mut libc::c_void }
+        } else {
+            Handler { _data: ptr::null_mut() }
+        }
     }
 
     pub unsafe fn drop_handler(handler: &mut Handler) {
-        munmap(handler._data, SIGSTKSZ);
+        if !handler._data.is_null() {
+            let stack =  libc::stack_t {
+                ss_sp: ptr::null_mut(),
+                ss_flags: SS_DISABLE,
+                ss_size: 0
+            };
+            sigaltstack(&stack, ptr::null_mut());
```

Comment:
This actually brings up an interesting race where if we try to disable the alternate stack while a thread is running on it, `sigaltstack` will return `EINVAL`. That should be pretty rare, though, but perhaps the `munmap` could only happen if this call succeeds?